src/vm/hints/hint_utils.rs

```diff
@@ -1315,6 +1341,98 @@ pub fn unsigned_div_rem(
     }
 }
 
+//  Implements hint:
+//  %{ vm_exit_scope() %}
+pub fn exit_scope(vm: &mut VirtualMachine) -> Result<(), VirtualMachineError> {
+    match vm.exec_scopes.exit_scope() {
+        Ok(_) => Ok(()),
+        Err(e) => Err(VirtualMachineError::MainScopeError(e)),
+    }
+}
+
+//  Implements hint:
+//  %{ vm_enter_scope({'n': ids.len}) %}
+pub fn memcpy_enter_scope(
+    vm: &mut VirtualMachine,
+    ids: HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let len_addr = get_address_from_var_name("len", ids, vm, hint_ap_tracking)?;
+
+    match vm.memory.get(&len_addr) {
+        Ok(Some(maybe_rel_len)) => {
+            let len = if let MaybeRelocatable::Int(len) = maybe_rel_len {
+                len
+            } else {
+                return Err(VirtualMachineError::ExpectedInteger(len_addr.clone()));
+            };
+            vm.exec_scopes.enter_scope(HashMap::from([(
+                String::from("n"),
+                PyValueType::BigInt(len.clone()),
+            )]));
+
+            Ok(())
+        }
+        _ => Err(VirtualMachineError::FailedToGetIds),
+    }
+}
+
+// Implements hint:
+// %{
+//     n -= 1
+//     ids.continue_copying = 1 if n > 0 else 0
+// %}
+pub fn memcpy_continue_copying(
+    vm: &mut VirtualMachine,
+    ids: HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let continue_copying_addr =
+        get_address_from_var_name("continue_copying", ids, vm, hint_ap_tracking)?;
+
+    // get `n` variable from vm scope
+    let mut n = match vm.exec_scopes.get_local_variables() {
+        Some(variables) => match variables.get("n") {
+            Some(PyValueType::BigInt(n)) => n.clone(),
+            _ => {
+                return Err(VirtualMachineError::VariableNotInScopeError(String::from(
+                    "n",
+                )))
+            }
+        },
+        None => return Err(VirtualMachineError::ScopeError),
+    };
+
+    // reassign `n` with `n - 1`
+    vm.exec_scopes
+        .assign_or_update_variable("n", PyValueType::BigInt(n - 1_i32));
+
+    // get new value of `n`
+    n = match vm.exec_scopes.get_local_variables() {
+        Some(variables) => match variables.get("n") {
+            Some(PyValueType::BigInt(n)) => n.clone(),
+            _ => {
+                return Err(VirtualMachineError::VariableNotInScopeError(String::from(
+                    "n",
+                )))
+            }
+        },
+        None => return Err(VirtualMachineError::ScopeError),
+    };
+
```

Comment:
Didn't you just write `n` in this scope? Why not reuse that value?

Reply:
```rust suggestion
     // reassign `n` with `n - 1`
    let new_n = n - 1_i32;
    vm.exec_scopes
        .assign_or_update_variable("n", PyValueType::BigInt(new_n.clone()));

    // if it is positive, insert 1 in the address of `continue_copying`
    // else, insert 0
    if new_n.is_positive() {
        vm.memory
            .insert(&continue_copying_addr, &MaybeRelocatable::Int(bigint!(1)))
            .map_err(VirtualMachineError::MemoryError)
    } else {
        vm.memory
            .insert(&continue_copying_addr, &MaybeRelocatable::Int(bigint!(0)))
            .map_err(VirtualMachineError::MemoryError)
    }
    };
```

should this be ok? I wanted to avoid the cloning of `new_n` but I can't seem to get it going with the borrow checker

Reply:
Is it possible to `.assign_or_update` at the end instead? That would allow moving `n` instead of cloning it.

Reply:
(Sorry, I meant `new_n`)

Reply:
good one! that worked nicely :)